src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HeadersUtilities.cs

```diff
@@ -0,0 +1,87 @@
+    using System.Globalization;
+    using System.Linq;
+
+    /// <summary>
+    /// Generic functions that can be used to get and set Http headers.
+    /// </summary>
+    internal static class HeadersUtilities
+    {
+        /// <summary>
+        /// Get the key value from the provided HttpHeader value that is set up as a comma-separated list of key value pairs. Each key value pair is formatted like (key)=(value).
+        /// </summary>
+        /// <param name="headerValues">The header values that may contain key name/value pairs.</param>
+        /// <param name="keyName">The name of the key value to find in the provided header values.</param>
+        /// <returns>The first key value, if it is found. If it is not found, then null.</returns>
+        public static string GetHeaderKeyValue(IEnumerable<string> headerValues, string keyName)
+        {
+            if (headerValues != null)
+            {
+                foreach (string keyNameValue in headerValues)
+                {
+                    string[] keyNameValueParts = keyNameValue.Trim().Split('=');
+                    if (keyNameValueParts.Length == 2 && keyNameValueParts[0].Trim() == keyName)
+                    {
+                        return keyNameValueParts[1].Trim();
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Given the provided list of header value strings, return a comma-separated list of key
+        /// name/value pairs with the provided keyName and keyValue. If the initial header value
+        /// strings contains the key name, then the original key value should be replaced with the
+        /// provided key value. If the initial header value strings don't contain the key name,
+        /// then the key name/value pair should be added to the comma-separated list and returned.
+        /// </summary>
+        /// <param name="headerValues">The existing header values that the key/value pair should be added to.</param>
+        /// <param name="keyName">The name of the key to add.</param>
+        /// <param name="keyValue">The value of the key to add.</param>
+        /// <returns>The result of setting the provided key name/value pair into the provided headerValues.</returns>
+        public static IEnumerable<string> SetHeaderKeyValue(IEnumerable<string> headerValues, string keyName, string keyValue)
+        {
+            string[] newHeaderKeyValue = new[] { string.Format(CultureInfo.InvariantCulture, "{0}={1}", keyName.Trim(), keyValue.Trim()) };
+            return headerValues == null || !headerValues.Any()
+                ? newHeaderKeyValue
+                : headerValues
+                    .Where((string headerValue) =>
+                    {
+                        int equalsSignIndex = headerValue.IndexOf('=');
+                        return equalsSignIndex == -1 || TrimSubstring(headerValue, 0, equalsSignIndex) != keyName;
```

Comment:
Avoid creating temporary strings just to do an equality check. Instead, do a custom comparison on the original string. (It's the same thing as doing ```s1.Equals(s2, StringComparison.InvariantCultureIgnoreCase)``` instead of ```s1.ToLowerInvariant().Equals(s2.ToLowerInvariant())```.)